File: web/tests/playwright/scripts/run-tests.ts
/**
 * Playwright Test Runner Script
 * Executes test suites with support for test dimension filtering.
 *
 * Converts test dimension flags (--coverage, --lens, etc.) into Playwright --grep patterns.
 * Example: --coverage smoke --path happy -> --grep "@coverage:smoke.*@path:happy"
 */

import {spawnSync} from "child_process"
import {existsSync, mkdirSync, readdirSync, writeFileSync} from "fs"
import {dirname, join, resolve} from "path"
import {fileURLToPath} from "url"

import {config as loadDotenv} from "dotenv"

const __dirname = dirname(fileURLToPath(import.meta.url))
// tests/ root is two levels up from tests/playwright/scripts/
const TESTS_ROOT = resolve(__dirname, "..", "..")
// web/ root is the workspace root holding both tests/ and packages/.
const WEB_ROOT = resolve(TESTS_ROOT, "..")

const TEST_LAYERS = ["unit", "integration", "acceptance"] as const
type TestLayer = (typeof TEST_LAYERS)[number]

// Each layer maps to a package script run across all workspace packages via
// `pnpm -r --if-present`. Packages that define the script (e.g. @agenta/entities)
// run their vitest suite for that layer; packages without it are skipped.
const LAYER_PACKAGE_SCRIPT: Record<TestLayer, string> = {
    unit: "test:unit",
    integration: "test:integration",
    acceptance: "test:acceptance",
}

// Test dimension types and their tag prefixes
const DIMENSION_PREFIXES: Record<string, string> = {
    coverage: "@coverage:",
    lens: "@lens:",
    path: "@path:",
    case: "@case:",
    speed: "@speed:",
    scope: "@scope:",
    license: "@license:",
    cost: "@cost:",
    plan: "@plan:",
    role: "@role:",
}

interface ParsedArgs {
    envFile: string | null
    layer: TestLayer | null
    grepPatterns: string[]
    playwrightArgs: string[]
}

function parseLayer(value: string): TestLayer {
    if (!(TEST_LAYERS as readonly string[]).includes(value)) {
        throw new Error(`Unsupported test layer: ${value} (expected ${TEST_LAYERS.join("|")})`)
    }
    return value as TestLayer
}

function parseArgs(args: string[]): ParsedArgs {
    let envFile: string | null = null
    let layer: TestLayer | null = null
    const grepPatterns: string[] = []
    const playwrightArgs: string[] = []

    let i = 0
    while (i < args.length) {
        const arg = args[i]

        // CLI separator from package managers (pnpm test -- --flag)
        if (arg === "--") {
            i++
            continue
        }

        if (arg === "--layer") {
            if (i + 1 >= args.length) {
                throw new Error("--layer requires a value")
            }
            layer = parseLayer(args[i + 1])
            i += 2
            continue
        }

        if (arg.startsWith("--layer=")) {
            layer = parseLayer(arg.slice("--layer=".length))
            i++
            continue
        }

        if (arg === "--env-file") {
            if (i + 1 >= args.length) {
                throw new Error("--env-file requires a value")
            }
            envFile = args[i + 1]
            i += 2
            continue
        }

        if (arg.startsWith("--env-file=")) {
            envFile = arg.slice("--env-file=".length)
            i++
            continue
        }

        // Check if this is a dimension flag
        const dimensionMatch = arg.match(
            /^--?(coverage|lens|path|case|speed|scope|license|cost|plan|role)$/,
        )

        if (dimensionMatch && i + 1 < args.length) {
            const dimension = dimensionMatch[1]
            const value = args[i + 1]
            const prefix = DIMENSION_PREFIXES[dimension]
            grepPatterns.push(`${prefix}${value}`)
            i += 2 // Skip both the flag and its value
        } else {
            // Pass through to playwright
            playwrightArgs.push(arg)
            i++
        }
    }

    return {envFile, layer, grepPatterns, playwrightArgs}
}

function getLicense(): string {
    return process.env.AGENTA_LICENSE === "ee" ? "ee" : "oss"
}

/** Mirrors run-tests.py: a layer folder counts as runnable only if it holds spec files. */
function hasSpecFiles(dir: string): boolean {
    if (!existsSync(dir)) {
        return false
    }
    for (const entry of readdirSync(dir, {withFileTypes: true})) {
        if (entry.isDirectory()) {
            if (hasSpecFiles(join(dir, entry.name))) {
                return true
            }
        } else if (entry.name.endsWith(".spec.ts") || entry.name.endsWith(".spec.tsx")) {
            return true
        }
    }
    return false
}

/** Write an empty JUnit report so the publish step has a file to consume on skipped layers. */
function writeEmptyResults(): void {
    const resultsDir = join(TESTS_ROOT, "results", getLicense())
    mkdirSync(resultsDir, {recursive: true})
    writeFileSync(
        join(resultsDir, "junit.xml"),
        '<?xml version="1.0" encoding="utf-8"?>' +
            '<testsuite tests="0" failures="0" errors="0" skipped="0"></testsuite>',
        "utf-8",
    )
}

function buildPlaywrightArgs(grepPatterns: string[], playwrightArgs: string[]): string[] {
    const args = ["test"]

    // Add grep pattern if we have dimension filters
    if (grepPatterns.length > 0) {
        // Combine patterns with .* to match all dimensions
        const grepExpression = grepPatterns.join(".*")
        args.push("--grep", grepExpression)
    }

    // Add remaining playwright arguments
    args.push(...playwrightArgs)

    return args
}

/**
 * Run the layer's vitest suite across every workspace package that defines the
 * matching `test:<layer>` script (via `pnpm -r --if-present`). Packages without
 * the script are skipped, so this is a no-op when no package has tests for the
 * layer. Returns the exit status (0 = success, including the no-package case).
 *
 * License, env vars, and markers are handled the same way as the Playwright
 * phase: AGENTA_LICENSE / AGENTA_TEST_LAYER and any --env-file values are
 * already on process.env and inherited here, and dimension markers are
 * forwarded to vitest's name filter (-t) using the same grep expression.
 */
function runVitestLayer(targetLayer: TestLayer, grepPatterns: string[]): number {
    const script = LAYER_PACKAGE_SCRIPT[targetLayer]
    const pnpmArgs = ["-r", "--if-present", "run", script]

    // Forward dimension markers to vitest as a name filter, mirroring how the
    // Playwright phase turns them into --grep. Args after `--` reach vitest.
    if (grepPatterns.length > 0) {
        pnpmArgs.push("--", "-t", grepPatterns.join(".*"))
    }

    console.log(`Executing: pnpm ${pnpmArgs.join(" ")} (AGENTA_LICENSE=${getLicense()})`)
    const result = spawnSync("pnpm", pnpmArgs, {
        cwd: WEB_ROOT,
        stdio: "inherit",
        shell: false,
        env: process.env,
    })
    return result.status ?? 1
}

// Parse command line arguments (skip node and script paths)
const args = process.argv.slice(2)
let parsedArgs: ParsedArgs
try {
    parsedArgs = parseArgs(args)
} catch (error) {
    console.error(error instanceof Error ? error.message : String(error))
    process.exit(1)
}

const {envFile, layer, grepPatterns, playwrightArgs} = parsedArgs

if (envFile) {
    const {error} = loadDotenv({path: envFile})
    if (error) {
        console.error(`Failed to load environment variables from ${envFile}: ${error.message}`)
        process.exit(1)
    }
    console.log(`Loaded environment variables from ${envFile}`)
}

// A layer runs every test source for that layer, regardless of runner:
//   1. package vitest suites (any package with a `test:<layer>` script), then
//   2. playwright specs in the license's `<license>/tests/playwright/<layer>`.
// Both phases share license/env/marker handling. We run both even if the first
// fails, then exit nonzero if either failed, so all failures surface in one run.
// Without --layer (raw playwright passthrough), only the playwright phase runs.
let failed = false

if (layer) {
    process.env.AGENTA_TEST_LAYER = layer

    // Phase 1 — package vitest suites (vitest runs before playwright).
    const vitestStatus = runVitestLayer(layer, grepPatterns)
    if (vitestStatus !== 0) {
        failed = true
    }

    // Phase 2 — playwright specs for this layer/license.
    const layerDir = join(TESTS_ROOT, "..", getLicense(), "tests", "playwright", layer)
    if (!hasSpecFiles(layerDir)) {
        console.log(
            `No ${layer} playwright specs for AGENTA_LICENSE=${getLicense()}; skipping playwright.`,
        )
        writeEmptyResults()
        process.exit(failed ? 1 : 0)
    }
    console.log(`Running ${layer} layer playwright (AGENTA_LICENSE=${getLicense()})`)
}

// Build and execute the playwright command
const playwrightCliArgs = buildPlaywrightArgs(grepPatterns, playwrightArgs)
console.log(`Executing: playwright ${playwrightCliArgs.join(" ")}`)

try {
    const result = spawnSync("playwright", playwrightCliArgs, {
        stdio: "inherit",
        shell: false,
    })
    if (result.status !== 0) {
        failed = true
    }
} catch {
    failed = true
}

process.exit(failed ? 1 : 0)

File: api/run-tests.py
#!/usr/bin/env python3
from typing import Optional
import os
import subprocess
import click

from dotenv import load_dotenv


ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
LOCAL_SDK_DIR = os.path.join(ROOT_DIR, "sdk")


TYPES = {
    "license": ["ee", "oss"],
    "layer": ["unit", "integration", "acceptance"],
    "coverage": ["smoke", "full"],
    "lens": ["functional", "performance", "security"],
    "plan": ["hobby", "pro", "business", "enterprise"],
    "role": ["owner", "admin", "developer", "editor", "annotator", "viewer"],
    "path": ["happy", "grumpy"],
    "case": ["typical", "edge"],
    "speed": ["fast", "slow"],
    "cost": ["free", "paid"],
}


def _has_pytest_option(pytest_args: Optional[tuple], option: str) -> bool:
    if not pytest_args:
        return False

    return any(arg == option or arg.startswith(f"{option}=") for arg in pytest_args)


def _resolve_license() -> str:
    return "ee" if os.getenv("AGENTA_LICENSE") == "ee" else "oss"


def _prepend_pythonpath(path: str) -> None:
    current = os.environ.get("PYTHONPATH")
    paths = [path]
    if current:
        paths.append(current)
    os.environ["PYTHONPATH"] = os.pathsep.join(paths)


def _has_test_files(path: str) -> bool:
    if not os.path.isdir(path):
        return False

    for root, _, files in os.walk(path):
        if "__pycache__" in root.split(os.sep):
            continue
        if any(
            file.endswith(".py")
            and (file.startswith("test_") or file.endswith("_test.py"))
            for file in files
        ):
            return True

    return False


def _resolve_test_dirs(license: str, layer: Optional[str]) -> list[str]:
    licenses = ["oss", "ee"] if license == "ee" else [license]
    test_dirs = []

    for license_name in licenses:
        path = os.path.join(license_name, "tests", "pytest")
        if layer:
            path = os.path.join(path, layer)
        if _has_test_files(path):
            test_dirs.append(path)

    return test_dirs


def _write_empty_results(results_dir: str) -> None:
    os.makedirs(results_dir, exist_ok=True)
    with open(os.path.join(results_dir, "junit.xml"), "w", encoding="utf-8") as junit:
        junit.write(
            '<?xml version="1.0" encoding="utf-8"?>'
            '<testsuite tests="0" failures="0" errors="0" skipped="0"></testsuite>'
        )
    with open(
        os.path.join(results_dir, "report.html"), "w", encoding="utf-8"
    ) as report:
        report.write("<html><body>No tests found.</body></html>")


@click.command()
@click.option(
    "--env-file",
    type=click.Path(exists=True, dir_okay=False),
    help="Path to a .env.* file with AGENTA_API_URL and AGENTA_AUTH_KEY",
)
@click.option(
    "--api-url",
    type=str,
    help="API URL for Agenta",
    envvar="AGENTA_API_URL",
)
@click.option(
    "--auth-key",
    type=str,
    help="Access token for Agenta",
    envvar="AGENTA_AUTH_KEY",
)
@click.option(
    "--coverage",
    type=click.Choice(TYPES["coverage"]),
    help="Coverage [smoke|full] (full = no coverage marker filter)",
    show_default=True,
)
@click.option(
    "--layer",
    type=click.Choice(TYPES["layer"]),
    help="Test layer [unit|integration|acceptance]",
)
@click.option(
    "--lens",
    type=click.Choice(TYPES["lens"]),
    help="Lens [functional|performance|security]",
    show_default=True,
)
@click.option(
    "--plan",
    type=click.Choice(TYPES["plan"]),
    help="Plan [hobby|pro|business|enterprise]",
)
@click.option(
    "--role",
    type=click.Choice(TYPES["role"]),
    help="Role [owner|admin|developer|editor|annotator|viewer]",
)
@click.option(
    "--path",
    type=click.Choice(TYPES["path"]),
    help="Path [happy|grumpy]",
)
@click.option(
    "--case",
    type=click.Choice(TYPES["case"]),
    help="Case [typical|edge]",
)
@click.option(
    "--speed",
    type=click.Choice(TYPES["speed"]),
    help="Speed [fast|slow]",
)
@click.option(
    "--cost",
    type=click.Choice(TYPES["cost"]),
    help="Cost [free|paid]",
)
@click.option(
    "--scope",
    help="Scope [...]",
)
@click.argument(
    "pytest_args",
    nargs=-1,
    type=click.UNPROCESSED,
)
def run_tests(
    env_file: Optional[str] = None,
    api_url: Optional[str] = None,
    auth_key: Optional[str] = None,
    layer: Optional[str] = None,
    coverage: Optional[str] = None,
    lens: Optional[str] = None,
    plan: Optional[str] = None,
    role: Optional[str] = None,
    path: Optional[str] = None,
    case: Optional[str] = None,
    speed: Optional[str] = None,
    cost: Optional[str] = None,
    scope: Optional[str] = None,
    pytest_args: Optional[tuple] = None,
):
    """
    Run pytest with dynamic markers and environment configuration.

    Additional args after '--' are passed directly to pytest.
    """
    marker_args = []

    if env_file:
        load_dotenv(env_file)
        click.echo(f"Loaded environment variables from {env_file}")
        if not api_url:
            api_url = os.getenv("AGENTA_API_URL")
        if not auth_key:
            auth_key = os.getenv("AGENTA_AUTH_KEY")

    # Set API_URL and AUTH_KEY as env vars for tests
    if api_url:
        os.environ["AGENTA_API_URL"] = api_url
        click.echo(f"AGENTA_API_URL={api_url}")
    if auth_key:
        os.environ["AGENTA_AUTH_KEY"] = auth_key
        L = len(auth_key)
        message = f"AGENTA_AUTH_KEY={auth_key[:2]}" + "." * (L - 4) + f"{auth_key[-2:]}"
        click.echo(message)

    license = _resolve_license()
    click.echo(f"AGENTA_LICENSE={license}")

    if os.path.isdir(LOCAL_SDK_DIR):
        _prepend_pythonpath(LOCAL_SDK_DIR)

    results_dir = os.path.join(license, "tests", "results")

    # Set optional dimensions
    for name, value in [
        ("COVERAGE", coverage),
        ("LENS", lens),
        ("PLAN", plan),
        ("ROLE", role),
        ("PATH", path),
        ("CASE", case),
        ("SPEED", speed),
        ("COST", cost),
        ("SCOPE", scope),
    ]:
        if value:
            if name == "COVERAGE" and value == "full":
                os.environ.pop("COVERAGE", None)
                click.echo("COVERAGE=full (coverage markers disabled)")
                continue
            os.environ[name] = value
            click.echo(f"{name}={value}")
            marker_args.append(f"{name.lower()}_{value}")

    # If pytest_args contains test paths, use them instead of test_dirs
    extra_paths = [a for a in (pytest_args or []) if not a.startswith("-")]
    test_dirs = _resolve_test_dirs(license, layer)
    if not extra_paths and not test_dirs:
        layer_label = f" {layer}" if layer else ""
        click.echo(
            f"No{layer_label} tests found for AGENTA_LICENSE={license}; skipping."
        )
        _write_empty_results(results_dir)
        return

    cmd = ["pytest"] + (extra_paths if extra_paths else test_dirs)

    if marker_args:
        marker_expr = " and ".join(marker_args)
        cmd += ["-m", marker_expr]

    os.makedirs(results_dir, exist_ok=True)

    if not _has_pytest_option(pytest_args, "--junit-xml"):
        cmd.append(f"--junit-xml={results_dir}/junit.xml")
    if not _has_pytest_option(pytest_args, "--html"):
        cmd.append(f"--html={results_dir}/report.html")

    if pytest_args:
        flags_only = [a for a in pytest_args if a.startswith("-")]
        cmd += flags_only

    click.echo(f"Executing: {' '.join(cmd)}")

    subprocess.run(cmd, check=True)


if __name__ == "__main__":
    run_tests()
